Add Search App documentation (#934)
saleor/saleor-docs: docs/developer/app-store/apps/search.mdx

---
title: Search
sidebar_position: 5
---

import { AppMetadata } from "/components/AppMetadata/AppMetadata.tsx";

<AppMetadata
  minSaleorVersion="3.10"
  roadmapUrl="https://github.com/saleor/apps/labels/App%3A%20Search"
  githubUrl="https://github.com/saleor/apps/tree/main/apps/search"
/>

## Introduction

The application provides integration with [Algolia](https://www.algolia.com/) search engine. It allows you to index your products and search them using Algolia's API.

## Features

- Synchronize the whole catalog with Algolia
- Update Algolia index on product changes

## Assumptions & limitations

> - Initial synchronization has to be done manually in the App dashboard

Reviewer: **polish**: I suggest dropping the "dashboard" bit, because it may be a bit confusing (app vs dashboard)

```suggestion
- Initial synchronization has to be done manually in the App
```

> - The application creates a separate record for each of the variants

Reviewer: **polish**:

```suggestion
- The application creates a separate record for each variant
```

- Each of the channels is indexed separately
- At this moment translations are not supported
- The exact state of the product quantity is not indexed, only the boolean value (in stock / out of stock)
- The app will synchronize product variants on the following [events](https://docs.saleor.io/docs/3.x/developer/extending/apps/asynchronous-webhooks):
  - `PRODUCT_CRETED`
  - `PRODUCT_UPDATED`
  - `PRODUCT_DELETED`
  - `PRODUCT_VARIANT_CREATED`
  - `PRODUCT_VARIANT_UPDATED`
  - `PRODUCT_VARIANT_DELETED`
  - `PRODUCT_VARIANT_OUT_OF_STOCK`
  - `PRODUCT_VARIANT_BACK_IN_STOCK`

## Glossary

- Algolia record: single entry in the Algolia index. Each Product Variant is a separate record.
- Index: a collection of Algolia records. We are creating separate indices for each of the channels.

## Permissions required by the app

- `MANAGE_PRODUCTS`

> - `MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES`

Reviewer 2: @krzysztofwolski lets add why these permissions are required. It will help to understand scope of the app. 

MANAGE_PRODUCTS is clear to me (but likely not for everyone), but I don't know why `MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES` is required

```mermaid
sequenceDiagram
    participant Saleor
    participant Search App
    participant Algolia API
    participant Storefront

    alt Initial synchronization
        Note over Search App: Staff user initializes synchronization in the Dashboard

        Search App->>+Saleor: Get Products data
        Saleor-->>-Search App: GraphQL response
        Search App->>Search App: Transform data to Algolia format
        Search App->>+Algolia API: Records with Product Variant data
        Algolia API-->>-Search App: Confirmation
    end

    alt Updating index on changes
        Note over Saleor: Products in the Saleor Instance changed

        Saleor->>+Search App: Webhook: product changed
        Search App->>Search App: Transform data to Algolia format
        Search App->>+Algolia API: Records with Product Variant data
        Algolia API-->>-Search App: Confirmation
        Search App-->>-Saleor: Index updated
    end

    alt Making queries to the Algolia
        Note over Storefront: User browse products catalog in the storefront

        Storefront->>+Algolia API: Product search request
        Algolia API-->>-Storefront: Search results
    end

```

## Pre-requisites

> - Account on [Algolia](https://www.algolia.com/). The free account will be sufficient, but has additional limitations, like maximum record size.

Reviewer: **polish**:

```suggestion
- [Algolia](https://www.algolia.com/) account. The free account will be sufficient, but has additional limitations, like maximum record size.
```

> - Saleor user with `MANAGE_APPS`, `MANAGE_PRODUCTS` and `MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES` permissions

Reviewer: **polish**:

```suggestion
- Saleor user with `MANAGE_APPS`, `MANAGE_PRODUCTS` and `MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES` permissions.
```

## Configuration

- Go to [API Keys](https://dashboard.algolia.com/account/api-keys/all) section in the Algolia dashboard to obtain API credentials
- Go to the Search App dashboard and fill required fields:
  - Application ID
  - Admin API Key
  - Index name prefix is optional, but recommended if you want to use the same Algolia account for multiple Saleor instances
- Click `Save`

> The application is now connected to your Algolia account. You can now synchronize your products:

Reviewer: **polish**: To get rid of one of the "now"s. 

```suggestion
The application is now connected to your Algolia account. You can start synchronizing your products:
```

- Click on the `Start importing` button in the `Index products` section
- Wait for the synchronization to finish.
- After the process is finished, apply recommended settings to the indices using the `Update indices configuration` button in the `Set indices settings` section

:::caution

Products synchronization **works on the frontend side**, which means the App **must remain open** during this operation.
In large shops, this operation can take a few seconds.

:::

> The application is now configured and synchronized, you can start making queries using the Algolia API or browse the data in the Algolia dashboard. Changes to the products will be automatically synchronized.

Reviewer: **polish:**

```suggestion
The application should now be configured and synchronized. You can start making queries using the Algolia API or browse the data in the Algolia dashboard. Changes to the products will be automatically synchronized.
```
